src/Microsoft.Restier.Publishers.OData/RestierController.cs

```diff
@@ -525,6 +578,127 @@ private ODataPath GetPath()
+                    var listResult = toListMethodInfo.Invoke(null, new object[] {bindingParameterValue});
+                    parameters.Add(listResult);
+                }
+                else
+                {
+                    parameters.Add(bindingParameterValue);
+                }
+            }
+            else
+            {
+                parameters.Add(bindingParameterValue.SingleOrDefault());
+            }
+
+            for (int i=1;i< parameterArray.Length;i++)
+            {
+                var parameter =parameterArray[i];
+                var currentParameterValue = segment.GetParameterValue(parameter.Name);
+                var parameterTypeRef = parameter.ParameterType.GetTypeReference(model);
+                var convertedValue = await DeserializationHelpers.ConvertValueAsync(currentParameterValue, parameter.ParameterType, parameterTypeRef, Api);
+                parameters.Add(convertedValue);
+            }
+
+            return await ExecuteFunction(method, parameters, cancellationToken);
+        }
+
+        private async Task<IQueryable> ExecuteFunction(MethodInfo method, Collection<object> parameters, CancellationToken cancellationToken)
+        {
+            object result = method.Invoke(Api, parameters.ToArray());
+            var model = await Api.GetModelAsync(cancellationToken);
+            IEdmTypeReference returnType = method.ReturnType.GetReturnTypeReference(model);
+            var asQueryableMethod = typeof(Queryable).GetMethod(
+                                        "AsQueryable",
+                                        BindingFlags.Static | BindingFlags.Public,
+                                        null,
+                                        new[] { typeof(IEnumerable<>) },
+                                        null);
+
+            if (returnType.IsCollection())
+            {
+                if (result == null)
+                {
+                    var elementClrType = method.ReturnType.GetElementType() ??
+                                         method.ReturnType.GenericTypeArguments[0];
+                    var constructor = typeof(List<>).MakeGenericType(elementClrType).GetConstructor(Type.EmptyTypes);
+                    var instance = constructor.Invoke(new object[] {});
+                    var emptyQuerable = asQueryableMethod.Invoke(null, new object[] { instance }) as IQueryable;
+                    return emptyQuerable;
+                }
+
+                var enumerableType = result.GetType().FindGenericType(typeof(IEnumerable<>));
+                if (enumerableType != null)
+                {
+                    return ((IEnumerable)result).AsQueryable();
+                }
+            }
+
+            // cannot return new[] { result }.AsQueryable(); as need to return in its own type but not object type
+            var objectQueryable = new[] { result }.AsQueryable();
```

Comment:
Hasn't quite got the point here...if control goes here, what does `result` look like? These code looks to me equivalent to `new[] {result}.AsQueryable().Cast<object>().AsQueryable()`...